# keyu chen
# km5ar

# Lab Assignment 8: Data Management Using `pandas`, Part 1
## DS 6001: Practice and Application of Data Science

### Instructions
Please answer the following questions as completely as possible using text, code, and the results of code as needed. Format your answers in a Jupyter notebook. To receive full credit, make sure you address every part of the problem, and make sure your document is formatted in a clean and professional way.

In this lab, you will be working with the [2017 Workplace Health in America survey](https://www.cdc.gov/workplacehealthpromotion/survey/data.html) which was conducted by the Centers for Disease Control and Prevention. According to the survey's [guidence document](https://www.cdc.gov/workplacehealthpromotion/data-surveillance/docs/2017-WHA-Guidance-Document-for-Use-of-Public-Data-files-508.pdf):

> The Workplace Health in America (WHA) Survey gathered information from a cross-sectional, nationally representative sample of US worksites. The sample was drawn from the Dun & Bradstreet (D&B) database of all private and public employers in the United States with at least 10 employees. Like previous national surveys, the worksite served as the sampling unit rather than the companies or firms to which the worksites belonged. Worksites were selected using a stratified simple random sample (SRS) design, where the primary strata were ten multi-state regions defined by the Centers for Disease Control and Prevention (CDC), plus an additional stratum containing all hospital worksites. 

The data contain over 300 features that report the industry and type of company where the respondents are employed, what kind of health insurance and other health programs are offered, and other characteristics of the workplaces including whether employees are allowed to work from home and the gender and age makeup of the workforce. The data are full of interesting information, but in order to make use of the data a great deal of data manipulation is required first.

## Problem 0
Import the following libraries:

In [1]:
!pip install sidetable==0.7.0
import numpy as np
import pandas as pd
import sidetable
import sqlite3
import warnings
warnings.filterwarnings('ignore')
import urllib

#pd.set_option('display.max_rows', 10) #display a maximumn of 10 rows
#pd.set_option('display.max_columns', None)

You should consider upgrading via the '/opt/venv/bin/python -m pip install --upgrade pip' command.


## Problem 1
The raw data are stored in an ASCII file on the 2017 Workplace Health in America survey [homepage](https://www.cdc.gov/workplacehealthpromotion/survey/data.html). Load the raw data directly into Python without downloading the data onto your harddrive and display a dataframe with only the 14th, 28th, and 102nd rows of the data. [1 point]

In [2]:
# https://www.cdc.gov/workplacehealthpromotion/data-surveillance/docs/whpps_120717.csv
url = "https://www.cdc.gov/workplacehealthpromotion/data-surveillance/docs/whpps_120717.csv"
df_1 = pd.read_csv(url, sep="~")
df_1

OC1  OC3  HI1  HI2  HI3  HI4  HRA1  HRA1A  HRA1B  HRA1E  ...  WL3_05  \
0      7  3.0  2.0  1.0  2.0  1.0   1.0    3.0    4.0    2.0  ...     PTO   
1      2  3.0  2.0  3.0  1.0  1.0   1.0    3.0    3.0    1.0  ...     NaN   
2      7  3.0  1.0  3.0  1.0  1.0   1.0    3.0   97.0    2.0  ...     NaN   
3      1  2.0  1.0  2.0  1.0  1.0  97.0   96.0   96.0   96.0  ...     NaN   
4      2  3.0  1.0  3.0  1.0  1.0   1.0    3.0    3.0    2.0  ...     NaN   
...   ..  ...  ...  ...  ...  ...   ...    ...    ...    ...  ...     ...   
2838   1  4.0  1.0  3.0  1.0  1.0   2.0   96.0   96.0   96.0  ...     NaN   
2839   7  4.0  2.0  3.0  1.0  1.0   1.0    1.0    1.0    2.0  ...     NaN   
2840   2  4.0  2.0  3.0  1.0  1.0   1.0    3.0    2.0    2.0  ...     NaN   
2841   1  4.0  2.0  3.0  1.0  1.0   1.0    2.0   97.0    2.0  ...     NaN   
2842   2  4.0  2.0  3.0  1.0  1.0   1.0    1.0    3.0    2.0  ...     NaN   

                  E1_09  Suppquex       Id  Region  CDC_Region  Industry  \
0                   NaN       2.0    217.0     1.0         2.0       7.0   
1                   NaN       1.0    326.0     3.0         7.0       7.0   
2                   NaN       1.0    399.0     4.0         8.0       7.0   
3                   NaN       1.0    475.0     5.0         9.0       7.0   
4                   NaN       1.0    489.0     2.0         4.0       7.0   
...                 ...       ...      ...     ...         ...       ...   
2838                NaN       2.0  29419.0     5.0        10.0       6.0   
2839                NaN       1.0  53041.0     5.0        10.0       6.0   
2840                NaN       1.0   7496.0     5.0        10.0       6.0   
2841                NaN       2.0  18994.0     5.0        10.0       6.0   
2842  Nothing is needed       2.0  27704.0     5.0        10.0       6.0   

      Size  Varstrata  Finalwt_worksite,,,,  
0      7.0        0.0      47.793940929,,,,  
1      6.0        0.0      47.793940929,,,,  
2      8.0        0.0      47.793940929,,,,  
3      4.0        0.0      47.793940929,,,,  
4      4.0        0.0      47.793940929,,,,  
...    ...        ...                   ...  
2838   5.0      549.0      421.28310196,,,,  
2839   5.0      549.0      421.28310196,,,,  
2840   8.0      552.0       62.86238594,,,,  
2841   8.0      552.0      76.460186994,,,,  
2842   8.0      552.0       62.86238594,,,,  

[2843 rows x 301 columns]

In [3]:
df_1_row_14_28_102 = df_1.iloc[[13,27,101],]
df_1_row_14_28_102

OC1  OC3  HI1  HI2  HI3  HI4  HRA1  HRA1A  HRA1B  HRA1E  ...  WL3_05  \
13    3  1.0  2.0  3.0  2.0  1.0   1.0    3.0    3.0    1.0  ...     NaN   
27    1  3.0  1.0  3.0  1.0  1.0   1.0    2.0    4.0    2.0  ...     NaN   
101   2  1.0  1.0  3.0  2.0  1.0   1.0    2.0    4.0    2.0  ...     NaN   

     E1_09  Suppquex       Id  Region  CDC_Region  Industry  Size  Varstrata  \
13     NaN       1.0   1437.0     4.0         6.0       7.0   3.0        0.0   
27     NaN       1.0   2501.0     2.0         4.0       7.0   8.0        0.0   
101    NaN       2.0  12636.0     4.0         6.0       7.0   4.0        0.0   

     Finalwt_worksite,,,,  
13       47.793940929,,,,  
27       47.793940929,,,,  
101      47.793940929,,,,  

[3 rows x 301 columns]

## Problem 2 
The data contain 301 columns. Create a new variable in Python's memory to store a working version of the data. In the working version, delete all of the columns except for the following:

* `Industry`: 7 Industry Categories with NAICS codes

* `Size`: 8 Employee Size Categories

* `OC3` Is your organization for profit, non-profit, government?

* `HI1` In general, do you offer full, partial or no payment of premiums for personal health insurance for full-time employees?

* `HI2` Over the past 12 months, were full-time employees asked to pay a larger proportion, smaller proportion or the same proportion of personal health insurance premiums?

* `HI3`: Does your organization offer personal health insurance for your part-time employees?

* `CP1`: Are there health education programs, which focus on skill development and lifestyle behavior change along with information dissemination and awareness building?

* `WL6`: Allow employees to work from home?

* Every column that begins `WD`, expressing the percentage of employees that have certain characteristics at the firm

[1 point]

In [4]:
df_2_a = df_1[['Industry','Size', 'OC3', 'HI1', 'HI2', 'HI3', 'CP1', 'WL6']]
df_2_b = df_1.loc[:,df_1.columns.str.startswith('WD')]

df_3 = pd.concat([df_2_a,df_2_b],axis = 1)
df_3

Industry  Size  OC3  HI1  HI2  HI3  CP1   WL6  WD1_1  WD1_2    WD2  \
0          7.0   7.0  3.0  2.0  1.0  2.0  1.0   1.0   25.0   20.0   85.0   
1          7.0   6.0  3.0  2.0  3.0  1.0  1.0   1.0  997.0  997.0   90.0   
2          7.0   8.0  3.0  1.0  3.0  1.0  1.0   1.0   35.0    4.0  997.0   
3          7.0   4.0  2.0  1.0  2.0  1.0  2.0   2.0   50.0   15.0   50.0   
4          7.0   4.0  3.0  1.0  3.0  1.0  1.0   1.0   50.0   40.0   60.0   
...        ...   ...  ...  ...  ...  ...  ...   ...    ...    ...    ...   
2838       6.0   5.0  4.0  1.0  3.0  1.0  1.0  99.0  999.0  999.0  999.0   
2839       6.0   5.0  4.0  2.0  3.0  1.0  1.0   2.0  997.0  997.0  997.0   
2840       6.0   8.0  4.0  2.0  3.0  1.0  1.0   1.0   27.0  997.0   61.0   
2841       6.0   8.0  4.0  2.0  3.0  1.0  2.0  99.0  999.0  999.0  999.0   
2842       6.0   8.0  4.0  2.0  3.0  1.0  1.0   1.0   25.0   10.0   35.0   

        WD3    WD4    WD5    WD6    WD7  
0      60.0   40.0   15.0    0.0   22.0  
1      90.0  997.0  997.0    0.0  997.0  
2     997.0   40.0   15.0  997.0  997.0  
3      85.0   75.0    0.0    0.0  997.0  
4      60.0   40.0   30.0    0.0   28.0  
...     ...    ...    ...    ...    ...  
2838  999.0  999.0  999.0  999.0  999.0  
2839  997.0  997.0  997.0  997.0  997.0  
2840  997.0  997.0  997.0  997.0  997.0  
2841  999.0  999.0  999.0  999.0  999.0  
2842   90.0   25.0   60.0   80.0  997.0  

[2843 rows x 16 columns]

## Problem 3
The [codebook](https://www.cdc.gov/workplacehealthpromotion/data-surveillance/docs/2017-WHA-Datafile-Codebook-508.pdf) for the WHA data contain short descriptions of the meaning of each of the columns in the data. Use these descriptions to decide on better and more intuitive names for the columns in the working version of the data, and rename the columns accordingly. [1 point]

In [5]:
df_3.rename(columns={'OC3': 'Organization_Category', 
'HI1': 'Insurance_offer_fulltime_employees', 
'HI2': 'Insurance_fulltime_proportion',
'HI3': 'Insurance_parttime',
'CP1': 'Health_Promoting', 
'WL6': 'work_home_availability',
'WD1_1':'percent_under30',
'WD1_2':'percent_over60',
'WD2': 'percent_female',
'WD3': 'percent_hourly',
'WD4': 'percent_other_daytime',
'WD5': 'percent_regularly_remotely/off_site',
'WD6': 'percent_bargain/unionized',
'WD7': 'average_annual_turnover'
}, inplace=True)

df_3

Industry  Size  Organization_Category  \
0          7.0   7.0                    3.0   
1          7.0   6.0                    3.0   
2          7.0   8.0                    3.0   
3          7.0   4.0                    2.0   
4          7.0   4.0                    3.0   
...        ...   ...                    ...   
2838       6.0   5.0                    4.0   
2839       6.0   5.0                    4.0   
2840       6.0   8.0                    4.0   
2841       6.0   8.0                    4.0   
2842       6.0   8.0                    4.0   

      Insurance_offer_fulltime_employees  Insurance_fulltime_proportion  \
0                                    2.0                            1.0   
1                                    2.0                            3.0   
2                                    1.0                            3.0   
3                                    1.0                            2.0   
4                                    1.0                            3.0   
...                                  ...                            ...   
2838                                 1.0                            3.0   
2839                                 2.0                            3.0   
2840                                 2.0                            3.0   
2841                                 2.0                            3.0   
2842                                 2.0                            3.0   

      Insurance_parttime  Health_Promoting  work_home_availability  \
0                    2.0               1.0                     1.0   
1                    1.0               1.0                     1.0   
2                    1.0               1.0                     1.0   
3                    1.0               2.0                     2.0   
4                    1.0               1.0                     1.0   
...                  ...               ...                     ...   
2838                 1.0               1.0                    99.0   
2839                 1.0               1.0                     2.0   
2840                 1.0               1.0                     1.0   
2841                 1.0               2.0                    99.0   
2842                 1.0               1.0                     1.0   

      percent_under30  percent_over60  percent_female  percent_hourly  \
0                25.0            20.0            85.0            60.0   
1               997.0           997.0            90.0            90.0   
2                35.0             4.0           997.0           997.0   
3                50.0            15.0            50.0            85.0   
4                50.0            40.0            60.0            60.0   
...               ...             ...             ...             ...   
2838            999.0           999.0           999.0           999.0   
2839            997.0           997.0           997.0           997.0   
2840             27.0           997.0            61.0           997.0   
2841            999.0           999.0           999.0           999.0   
2842             25.0            10.0            35.0            90.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      40.0                                 15.0   
1                     997.0                                997.0   
2                      40.0                                 15.0   
3                      75.0                                  0.0   
4                      40.0                                 30.0   
...                     ...                                  ...   
2838                  999.0                                999.0   
2839                  997.0                                997.0   
2840                  997.0                                997.0   
2841                  999.0                                999.0   
2842                   25.0                                 60.0   

     

HI1: Health Insurance HI1: In general, do you offer full, partial or no payment of premiums for personal health insurance for full-time employees?

HI2: Health Insurance HI2: Over the past 12 months, were full-time employees asked to pay a larger proportion, smaller proportion or the same proportion of personal health insurance premiums?

HI3: Health Insurance HI3: Does your organization offer personal health insurance for your part-time employees?

CP1: Promoting Health at Work CP1: Health education programs, which focus on skill development and lifestyle behavior change along with information dissemination and awareness building?

WL6: Work-Life Programming, Policies and Benefits WL6: Allow employees to work from home?

Industry: 7 Industry Categories

Size: 8 Employee Size Categories

WD1_1: Workforce Demographics WD1_1: What percentage of employees are under age 30?

WD1_2: Workforce Demographics WD1_2: What percentage of employees are age 60 or older?

WD2: Workforce Demographics WD2: What percentage of employees are female?

WD3: Workforce Demographics WD3: What percentage of employees are hourly/non-exempt?

WD4: Workforce Demographics WD4: What percentage of employees work something other than a typical daytime shift?

WD5: Workforce Demographics WD5: What percentage of employees regularly work remotely or off-site?

WD6: Workforce Demographics WD6: What percentage of employees are under collective bargaining or unionized?

WD7: Workforce Demographics WD7: What is the average annual percentage of turnover at this worksite location?


## Problem 4
Using the codebook and this [dictionary of NAICS industrial codes](https://www.naics.com/search-naics-codes-by-industry/), place descriptive labels on the categories of the industry column in the working data. [1 point]

Industry Column:

Category 1: 
NAICS Sectors: Agriculture, Forestry, Fishing and Hunting, Mining, Utilities, Construction, Manufacturing

Category 2: 
NAICS Sectors: 	Wholesale Trade, Retail Trade, Transportation and Warehousing

Category 3: 
NAICS Sectors: 	Arts, Entertainment, and Recreation, Accommodation and Food Services, Other Services (except Public Administration)

Category 4: 
NAICS Sectors: Information, Finance and Insurance, Real Estate Rental and Leasing, Professional, Scientific, and Technical Services, 
               Management of Companies and Enterprises, Administrative and Support and Waste Management and Remediation Services

Category 5: 
NAICS Sectors: Educational Services, Health Care and Social Assistance(excluding hospital worksites)

Category 6: 
NAICS Sectors: Public Administration

Category 7: 
NAICS Sectors: General Medical and Surgical Hospitals, Psychiatric and Substance Abuse Hospitals, 
               Specialty (except Psychiatric and Substance Abuse) Hospitals

In [6]:
df_3.Industry.value_counts()

5.0    551
1.0    525
3.0    433
4.0    429
7.0    338
2.0    311
6.0    255
Name: Industry, dtype: int64

In [7]:
df_4 = df_3.copy(deep=True)
replace_map = {1:'Agriculture',
               2:'WholesaleTrade', 
               3:'Arts_Entertainment', 
               4:'Information_Finance',
               5:'EducationalServices',
               6:'PublicAdministration',
               7:'Hospitals'
               }
df_4.Industry = df_4.Industry.map(replace_map)

df_4.Industry
df_4

Industry  Size  Organization_Category  \
0                Hospitals   7.0                    3.0   
1                Hospitals   6.0                    3.0   
2                Hospitals   8.0                    3.0   
3                Hospitals   4.0                    2.0   
4                Hospitals   4.0                    3.0   
...                    ...   ...                    ...   
2838  PublicAdministration   5.0                    4.0   
2839  PublicAdministration   5.0                    4.0   
2840  PublicAdministration   8.0                    4.0   
2841  PublicAdministration   8.0                    4.0   
2842  PublicAdministration   8.0                    4.0   

      Insurance_offer_fulltime_employees  Insurance_fulltime_proportion  \
0                                    2.0                            1.0   
1                                    2.0                            3.0   
2                                    1.0                            3.0   
3                                    1.0                            2.0   
4                                    1.0                            3.0   
...                                  ...                            ...   
2838                                 1.0                            3.0   
2839                                 2.0                            3.0   
2840                                 2.0                            3.0   
2841                                 2.0                            3.0   
2842                                 2.0                            3.0   

      Insurance_parttime  Health_Promoting  work_home_availability  \
0                    2.0               1.0                     1.0   
1                    1.0               1.0                     1.0   
2                    1.0               1.0                     1.0   
3                    1.0               2.0                     2.0   
4                    1.0               1.0                     1.0   
...                  ...               ...                     ...   
2838                 1.0               1.0                    99.0   
2839                 1.0               1.0                     2.0   
2840                 1.0               1.0                     1.0   
2841                 1.0               2.0                    99.0   
2842                 1.0               1.0                     1.0   

      percent_under30  percent_over60  percent_female  percent_hourly  \
0                25.0            20.0            85.0            60.0   
1               997.0           997.0            90.0            90.0   
2                35.0             4.0           997.0           997.0   
3                50.0            15.0            50.0            85.0   
4                50.0            40.0            60.0            60.0   
...               ...             ...             ...             ...   
2838            999.0           999.0           999.0           999.0   
2839            997.0           997.0           997.0           997.0   
2840             27.0           997.0            61.0           997.0   
2841            999.0           999.0           999.0           999.0   
2842             25.0            10.0            35.0            90.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      40.0                                 15.0   
1                     997.0                                997.0   
2                      40.0                                 15.0   
3                      75.0                                  0.0   
4                      40.0                                 30.0   
...                     ...                                  ...   
2838                  999.0                                999.0   
2839                  997.0                                997.0   
2840                  997.0                                997.0   
2841      

## Problem 5
Using the codebook, recode the "size" column to have three categories: "Small" for workplaces with fewer than 100 employees, "Medium" for workplaces with at least 100 but fewer than 500 employees, and "Large" for companies with at least 500 employees. [Note: Python dataframes have an attribute `.size` that reports the space the dataframe takes up in memory. Don't confuse this attribute with the column named "Size" in the raw data.] [1 point]

In [8]:
df_5 = df_4.copy(deep=True)
replace_map2 = {1:'Small', 2:'Small', 3:'Small', 4:'Medium', 
               5:'Medium', 6:'Large', 7:'Large', 8:'Large'}
df_5.Size = df_5.Size.map(replace_map2)
df_5

Industry    Size  Organization_Category  \
0                Hospitals   Large                    3.0   
1                Hospitals   Large                    3.0   
2                Hospitals   Large                    3.0   
3                Hospitals  Medium                    2.0   
4                Hospitals  Medium                    3.0   
...                    ...     ...                    ...   
2838  PublicAdministration  Medium                    4.0   
2839  PublicAdministration  Medium                    4.0   
2840  PublicAdministration   Large                    4.0   
2841  PublicAdministration   Large                    4.0   
2842  PublicAdministration   Large                    4.0   

      Insurance_offer_fulltime_employees  Insurance_fulltime_proportion  \
0                                    2.0                            1.0   
1                                    2.0                            3.0   
2                                    1.0                            3.0   
3                                    1.0                            2.0   
4                                    1.0                            3.0   
...                                  ...                            ...   
2838                                 1.0                            3.0   
2839                                 2.0                            3.0   
2840                                 2.0                            3.0   
2841                                 2.0                            3.0   
2842                                 2.0                            3.0   

      Insurance_parttime  Health_Promoting  work_home_availability  \
0                    2.0               1.0                     1.0   
1                    1.0               1.0                     1.0   
2                    1.0               1.0                     1.0   
3                    1.0               2.0                     2.0   
4                    1.0               1.0                     1.0   
...                  ...               ...                     ...   
2838                 1.0               1.0                    99.0   
2839                 1.0               1.0                     2.0   
2840                 1.0               1.0                     1.0   
2841                 1.0               2.0                    99.0   
2842                 1.0               1.0                     1.0   

      percent_under30  percent_over60  percent_female  percent_hourly  \
0                25.0            20.0            85.0            60.0   
1               997.0           997.0            90.0            90.0   
2                35.0             4.0           997.0           997.0   
3                50.0            15.0            50.0            85.0   
4                50.0            40.0            60.0            60.0   
...               ...             ...             ...             ...   
2838            999.0           999.0           999.0           999.0   
2839            997.0           997.0           997.0           997.0   
2840             27.0           997.0            61.0           997.0   
2841            999.0           999.0           999.0           999.0   
2842             25.0            10.0            35.0            90.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      40.0                                 15.0   
1                     997.0                                997.0   
2                      40.0                                 15.0   
3                      75.0                                  0.0   
4                      40.0                                 30.0   
...                     ...                                  ...   
2838                  999.0                                999.0   
2839                  997.0                                997.0   
2840                  997.0                           

## Problem 6
Use the codebook to write accurate and descriptive labels for each category for each categorical column in the working data. Then apply all of these labels to the data at once. Code "Legitimate Skip", "Don't know", "Refused", and "Blank" as missing values. [2 points]

In [9]:
df_6 = df_5.copy(deep=True)

In [10]:
replace_map = {'Organization_Category':{1:'profit_public',
                2:'profit_private', 
                3:'Non_profit', 
                4:'state_or_local_gov',
                5:'federal_gov',
                6:'other',
                97: np.nan,
                98: np.nan,
                99: np.nan},
              'Insurance_offer_fulltime_employees':{1:'Full',
                2:'Partial', 
                3:'None', 
                97: np.nan,
                98: np.nan,
                99: np.nan},
              'Insurance_fulltime_proportion':{1:'Larger',
                2:'Smaller', 
                3:'Same',
                96: np.nan, 
                97: np.nan,
                98: np.nan,
                99: np.nan},
              'Insurance_parttime':{1:'Yes',
                2:'No', 
                97: np.nan,
                98: np.nan,
                99: np.nan},
              'Health_Promoting':{1:'Yes',
                2:'No', 
                97: np.nan,
                98: np.nan},
              'work_home_availability':{1:'Yes',
                2:'No', 
                97: np.nan,
                98: np.nan,
                99: np.nan}}
#anes_clean = anes_clean.replace(replace_map)

In [11]:
df_6 = df_6.replace(replace_map)

In [12]:
# test code
df_6[['Organization_Category','Insurance_offer_fulltime_employees','Insurance_fulltime_proportion','Insurance_parttime','Health_Promoting','work_home_availability']]

Organization_Category Insurance_offer_fulltime_employees  \
0               Non_profit                            Partial   
1               Non_profit                            Partial   
2               Non_profit                               Full   
3           profit_private                               Full   
4               Non_profit                               Full   
...                    ...                                ...   
2838    state_or_local_gov                               Full   
2839    state_or_local_gov                            Partial   
2840    state_or_local_gov                            Partial   
2841    state_or_local_gov                            Partial   
2842    state_or_local_gov                            Partial   

     Insurance_fulltime_proportion Insurance_parttime Health_Promoting  \
0                           Larger                 No              Yes   
1                             Same                Yes              Yes   
2                             Same                Yes              Yes   
3                          Smaller                Yes               No   
4                             Same                Yes              Yes   
...                            ...                ...              ...   
2838                          Same                Yes              Yes   
2839                          Same                Yes              Yes   
2840                          Same                Yes              Yes   
2841                          Same                Yes               No   
2842                          Same                Yes              Yes   

     work_home_availability  
0                       Yes  
1                       Yes  
2                       Yes  
3                        No  
4                       Yes  
...                     ...  
2838                    NaN  
2839                     No  
2840                    Yes  
2841                    NaN  
2842                    Yes  

[2843 rows x 6 columns]

## Problem 7
The features that measure the percent of the workforce with a particular characteristic use the codes 997, 998, and 999 to represent "Don't know", "Refusal", and "Blank/Invalid" respectively. Replace these values with missing values for all of the percentage features at the same time. [1 point]

In [13]:
df_7 = df_6.copy(deep=True)

In [14]:
replace_map_2 = {'percent_under30':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_over60':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_female':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_hourly':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_other_daytime':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_regularly_remotely/off_site':{997: np.nan,
                998: np.nan,
                999: np.nan},
              'percent_bargain/unionized':{997: np.nan,
                998: np.nan,
                999: np.nan},
                'average_annual_turnover':{997: np.nan,
                998: np.nan,
                999: np.nan}}

In [15]:
df_7 = df_7.replace(replace_map_2)

In [16]:
# test
df_7.loc[:,'percent_under30':'average_annual_turnover']

percent_under30  percent_over60  percent_female  percent_hourly  \
0                25.0            20.0            85.0            60.0   
1                 NaN             NaN            90.0            90.0   
2                35.0             4.0             NaN             NaN   
3                50.0            15.0            50.0            85.0   
4                50.0            40.0            60.0            60.0   
...               ...             ...             ...             ...   
2838              NaN             NaN             NaN             NaN   
2839              NaN             NaN             NaN             NaN   
2840             27.0             NaN            61.0             NaN   
2841              NaN             NaN             NaN             NaN   
2842             25.0            10.0            35.0            90.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      40.0                                 15.0   
1                       NaN                                  NaN   
2                      40.0                                 15.0   
3                      75.0                                  0.0   
4                      40.0                                 30.0   
...                     ...                                  ...   
2838                    NaN                                  NaN   
2839                    NaN                                  NaN   
2840                    NaN                                  NaN   
2841                    NaN                                  NaN   
2842                   25.0                                 60.0   

      percent_bargain/unionized  average_annual_turnover  
0                           0.0                     22.0  
1                           0.0                      NaN  
2                           NaN                      NaN  
3                           0.0                      NaN  
4                           0.0                     28.0  
...                         ...                      ...  
2838                        NaN                      NaN  
2839                        NaN                      NaN  
2840                        NaN                      NaN  
2841                        NaN                      NaN  
2842                       80.0                      NaN  

[2843 rows x 8 columns]

## Problem 8
Sort the working data by industry in ascending alphabetical order. Within industry categories, sort the rows by size in ascending alphabetical order. Within groups with the same industry and size, sort by percent of the workforce that is under 30 in descending numeric order. [1 point]

In [17]:
df_8 = df_7.copy(deep=True)

In [18]:
df_8 = df_8.sort_values(by = ['Industry','Size','percent_under30'], ascending= [True, True, False])
df_8

Industry   Size Organization_Category  \
1732     Agriculture  Large        profit_private   
1476     Agriculture  Large        profit_private   
1477     Agriculture  Large        profit_private   
704      Agriculture  Large        profit_private   
1241     Agriculture  Large        profit_private   
...              ...    ...                   ...   
2604  WholesaleTrade  Small            Non_profit   
2626  WholesaleTrade  Small        profit_private   
2629  WholesaleTrade  Small         profit_public   
2631  WholesaleTrade  Small        profit_private   
1662             NaN    NaN                   NaN   

     Insurance_offer_fulltime_employees Insurance_fulltime_proportion  \
1732                            Partial                          Same   
1476                            Partial                          Same   
1477                            Partial                       Smaller   
704                                Full                          Same   
1241                               Full                          Same   
...                                 ...                           ...   
2604                               Full                          Same   
2626                            Partial                        Larger   
2629                               Full                        Larger   
2631                            Partial                        Larger   
1662                                NaN                           NaN   

     Insurance_parttime Health_Promoting work_home_availability  \
1732                 No              Yes                     No   
1476                 No              Yes                     No   
1477                 No              Yes                    Yes   
704                  No              Yes                    Yes   
1241                 No              Yes                    Yes   
...                 ...              ...                    ...   
2604                 No              Yes                     No   
2626                Yes               No                     No   
2629                 No              Yes                    Yes   
2631                Yes               No                     No   
1662                NaN              NaN                    NaN   

      percent_under30  percent_over60  percent_female  percent_hourly  \
1732             50.0            10.0            50.0            75.0   
1476             40.0            10.0            30.0            60.0   
1477             25.0            15.0            20.0            60.0   
704              20.0            15.0            17.0            62.0   
1241             20.0            25.0            50.0            70.0   
...               ...             ...             ...             ...   
2604              NaN             NaN             NaN             NaN   
2626              NaN             NaN             NaN             NaN   
2629              NaN             2.0            15.0             NaN   
2631              NaN             NaN             NaN            95.0   
1662              NaN             NaN             NaN             NaN   

      percent_other_daytime  percent_regularly_remotely/off_site  \
1732                   10.0                                  0.0   
1476                   30.0                                  5.0   
1477                   10.0                                  2.0   
704                    10.0                                  5.0   
1241                   20.0                                  5.0   
...                     ...                                  ...   
2604                    NaN                                  NaN   
2626                    NaN                                  NaN   
2629                    NaN                                 90.0   
2631                    NaN                                  NaN   
1662                    NaN                                  NaN   


## Problem 9
There is one row in the working data that has a `NaN` value for industry. Delete this row. Use a logical expression, and not the row number. [1 point]

In [19]:
df_9 = df_8.copy(deep=True)

In [20]:
df_9.dropna(subset=['Industry'], inplace=True)
df_9.reset_index(drop=True, inplace=True)
df_9

Industry   Size Organization_Category  \
0        Agriculture  Large        profit_private   
1        Agriculture  Large        profit_private   
2        Agriculture  Large        profit_private   
3        Agriculture  Large        profit_private   
4        Agriculture  Large        profit_private   
...              ...    ...                   ...   
2837  WholesaleTrade  Small        profit_private   
2838  WholesaleTrade  Small            Non_profit   
2839  WholesaleTrade  Small        profit_private   
2840  WholesaleTrade  Small         profit_public   
2841  WholesaleTrade  Small        profit_private   

     Insurance_offer_fulltime_employees Insurance_fulltime_proportion  \
0                               Partial                          Same   
1                               Partial                          Same   
2                               Partial                       Smaller   
3                                  Full                          Same   
4                                  Full                          Same   
...                                 ...                           ...   
2837                               Full                          Same   
2838                               Full                          Same   
2839                            Partial                        Larger   
2840                               Full                        Larger   
2841                            Partial                        Larger   

     Insurance_parttime Health_Promoting work_home_availability  \
0                    No              Yes                     No   
1                    No              Yes                     No   
2                    No              Yes                    Yes   
3                    No              Yes                    Yes   
4                    No              Yes                    Yes   
...                 ...              ...                    ...   
2837                Yes               No                     No   
2838                 No              Yes                     No   
2839                Yes               No                     No   
2840                 No              Yes                    Yes   
2841                Yes               No                     No   

      percent_under30  percent_over60  percent_female  percent_hourly  \
0                50.0            10.0            50.0            75.0   
1                40.0            10.0            30.0            60.0   
2                25.0            15.0            20.0            60.0   
3                20.0            15.0            17.0            62.0   
4                20.0            25.0            50.0            70.0   
...               ...             ...             ...             ...   
2837              NaN             2.0             2.0            90.0   
2838              NaN             NaN             NaN             NaN   
2839              NaN             NaN             NaN             NaN   
2840              NaN             2.0            15.0             NaN   
2841              NaN             NaN             NaN            95.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      10.0                                  0.0   
1                      30.0                                  5.0   
2                      10.0                                  2.0   
3                      10.0                                  5.0   
4                      20.0                                  5.0   
...                     ...                                  ...   
2837                    0.0                                  0.0   
2838                    NaN                                  NaN   
2839                    NaN                                  NaN   
2840                    NaN                                 90.0   
2841                    NaN                                  NaN   


## Problem 10
Create a new feature named `gender_balance` that has three categories: "Mostly men" for workplaces with between 0% and 35% female employees, "Balanced" for workplaces with more than 35% and at most 65% female employees, and "Mostly women" for workplaces with more than 65% female employees. [1 point]

In [21]:
df_10 = df_9.copy(deep=True)

In [22]:
df_10['gender_balance'] = pd.cut(df_10.percent_female, bins=[-0.1,35,65,100], labels=("Mostly men", "Balanced", "Mostly women"))
df_10[['gender_balance','percent_female']]

gender_balance  percent_female
0          Balanced            50.0
1        Mostly men            30.0
2        Mostly men            20.0
3        Mostly men            17.0
4          Balanced            50.0
...             ...             ...
2837     Mostly men             2.0
2838            NaN             NaN
2839            NaN             NaN
2840     Mostly men            15.0
2841            NaN             NaN

[2842 rows x 2 columns]

## Problem 11
Change the data type of all categorical features in the working data from "object" to "category". [1 point]

In [23]:
df_11 = df_10.copy(deep=True)
df_11

Industry   Size Organization_Category  \
0        Agriculture  Large        profit_private   
1        Agriculture  Large        profit_private   
2        Agriculture  Large        profit_private   
3        Agriculture  Large        profit_private   
4        Agriculture  Large        profit_private   
...              ...    ...                   ...   
2837  WholesaleTrade  Small        profit_private   
2838  WholesaleTrade  Small            Non_profit   
2839  WholesaleTrade  Small        profit_private   
2840  WholesaleTrade  Small         profit_public   
2841  WholesaleTrade  Small        profit_private   

     Insurance_offer_fulltime_employees Insurance_fulltime_proportion  \
0                               Partial                          Same   
1                               Partial                          Same   
2                               Partial                       Smaller   
3                                  Full                          Same   
4                                  Full                          Same   
...                                 ...                           ...   
2837                               Full                          Same   
2838                               Full                          Same   
2839                            Partial                        Larger   
2840                               Full                        Larger   
2841                            Partial                        Larger   

     Insurance_parttime Health_Promoting work_home_availability  \
0                    No              Yes                     No   
1                    No              Yes                     No   
2                    No              Yes                    Yes   
3                    No              Yes                    Yes   
4                    No              Yes                    Yes   
...                 ...              ...                    ...   
2837                Yes               No                     No   
2838                 No              Yes                     No   
2839                Yes               No                     No   
2840                 No              Yes                    Yes   
2841                Yes               No                     No   

      percent_under30  percent_over60  percent_female  percent_hourly  \
0                50.0            10.0            50.0            75.0   
1                40.0            10.0            30.0            60.0   
2                25.0            15.0            20.0            60.0   
3                20.0            15.0            17.0            62.0   
4                20.0            25.0            50.0            70.0   
...               ...             ...             ...             ...   
2837              NaN             2.0             2.0            90.0   
2838              NaN             NaN             NaN             NaN   
2839              NaN             NaN             NaN             NaN   
2840              NaN             2.0            15.0             NaN   
2841              NaN             NaN             NaN            95.0   

      percent_other_daytime  percent_regularly_remotely/off_site  \
0                      10.0                                  0.0   
1                      30.0                                  5.0   
2                      10.0                                  2.0   
3                      10.0                                  5.0   
4                      20.0                                  5.0   
...                     ...                                  ...   
2837                    0.0                                  0.0   
2838                    NaN                                  NaN   
2839                    NaN                                  NaN   
2840                    NaN                                 90.0   
2841                    NaN                                  NaN   


In [24]:
category_cols = ['Industry', 'Size', 'Organization_Category', 'Insurance_offer_fulltime_employees', 'Insurance_fulltime_proportion', 'Insurance_parttime', 'Health_Promoting','work_home_availability', 'gender_balance']
df_11[category_cols] = df_11[category_cols].astype('category')
df_11.dtypes

Industry                               category
Size                                   category
Organization_Category                  category
Insurance_offer_fulltime_employees     category
Insurance_fulltime_proportion          category
Insurance_parttime                     category
Health_Promoting                       category
work_home_availability                 category
percent_under30                         float64
percent_over60                          float64
percent_female                          float64
percent_hourly                          float64
percent_other_daytime                   float64
percent_regularly_remotely/off_site     float64
percent_bargain/unionized               float64
average_annual_turnover                 float64
gender_balance                         category
dtype: object

## Problem 12
Filter the data to only those rows that represent small workplaces that allow employees to work from home. Then report how many of these workplaces offer full insurance, partial insurance, and no insurance. Use a function that reports the percent, cumulative count, and cumulative percent in addition to the counts. [1 point]

In [25]:
df_12 = df_11.copy(deep=True)

In [26]:
df_12_b = df_12.query("work_home_availability == 'Yes' & Size == 'Small'")
df_12_b

Industry   Size Organization_Category  \
74       Agriculture  Small                   NaN   
76       Agriculture  Small        profit_private   
77       Agriculture  Small        profit_private   
80       Agriculture  Small        profit_private   
83       Agriculture  Small        profit_private   
...              ...    ...                   ...   
2823  WholesaleTrade  Small                 other   
2826  WholesaleTrade  Small                   NaN   
2828  WholesaleTrade  Small                 other   
2832  WholesaleTrade  Small                   NaN   
2840  WholesaleTrade  Small         profit_public   

     Insurance_offer_fulltime_employees Insurance_fulltime_proportion  \
74                                 None                           NaN   
76                                 Full                          Same   
77                              Partial                          Same   
80                                 Full                          Same   
83                              Partial                          Same   
...                                 ...                           ...   
2823                            Partial                           NaN   
2826                            Partial                       Smaller   
2828                               Full                          Same   
2832                                NaN                           NaN   
2840                               Full                        Larger   

     Insurance_parttime Health_Promoting work_home_availability  \
74                   No               No                    Yes   
76                   No               No                    Yes   
77                   No               No                    Yes   
80                   No               No                    Yes   
83                   No               No                    Yes   
...                 ...              ...                    ...   
2823                Yes              NaN                    Yes   
2826                 No               No                    Yes   
2828                 No               No                    Yes   
2832                Yes              Yes                    Yes   
2840                 No              Yes                    Yes   

      percent_under30  percent_over60  percent_female  percent_hourly  \
74              100.0             0.0            90.0            90.0   
76               95.0             1.0             1.0            50.0   
77               90.0            20.0             NaN             NaN   
80               75.0             1.0             5.0             5.0   
83               75.0             4.0             1.0            75.0   
...               ...             ...             ...             ...   
2823              NaN             NaN            15.0            80.0   
2826              NaN             NaN             NaN             NaN   
2828              NaN             NaN             NaN             NaN   
2832              NaN             NaN             NaN            20.0   
2840              NaN             2.0            15.0             NaN   

      percent_other_daytime  percent_regularly_remotely/off_site  \
74                      1.0                                  1.0   
76                     50.0                                  5.0   
77                      NaN                                  NaN   
80                      0.0                                 75.0   
83                     30.0                                  5.0   
...                     ...                                  ...   
2823                    NaN                                  NaN   
2826                    NaN                                  NaN   
2828                    0.0                                  5.0   
2832                    NaN                                  NaN   
2840                    NaN                                 90.0   


In [27]:
df_12_b.stb.freq(['Insurance_offer_fulltime_employees'])

Insurance_offer_fulltime_employees  count    percent  cumulative_count  \
0                               Full    324  46.285714               324   
1                            Partial    310  44.285714               634   
2                               None     66   9.428571               700   

   cumulative_percent  
0           46.285714  
1           90.571429  
2          100.000000

## Problem 13
Anything that can be done in SQL can be done with `pandas`. The next several questions ask you to write `pandas` code to match a given SQL query. But to check that the SQL query and `pandas` code yield the same result, create a new database wsing the `sqlite3` package and input the cleaned WHA data as a table in this database. (See module 6 for a discussion of SQlite in Python.) [1 point]

In [28]:
#!pip install wget
#!pip install sqlite3
#!pip install sqlalchemy
#!pip install dotenv

In [29]:
import numpy as np
import pandas as pd
import wget
import sqlite3
import sqlalchemy
import requests
import json
import os
import sys
#import dotenv
#os.chdir("/Users/jk8sd/Box Sync/Practice and Applications 1 online/Module 6 - Building and Connecting to Databases") # change to the directory where your .env file is
#os.chdir("/Users/cory/Desktop/DS 6001/lab_4") 
#dotenv.load_dotenv() # register the .env file where passwords are stored
sys.tracebacklimit = 0 # turn off the error tracebacks

In [30]:
import sqlite3
whpps = sqlite3.connect("whpps.db") 

In [31]:
df_final = df_12.copy(deep=True)

In [32]:
df_final.to_sql('whpps', whpps, index=False, chunksize=1000, if_exists='replace')

In [33]:
whpps_cursor = whpps.cursor()
whpps_cursor

## Problem 14
Write `pandas` code that replicates the output of the following SQL code:
```
SELECT size, type, premiums AS insurance, percent_female FROM whpps
WHERE industry = 'Hospitals' AND premium_change='Smaller'
ORDER BY percent_female DESC;
```
For each of these queries, your feature names might be different from the ones listed in the query, depending on the names you chose in problem 3.
[2 points]

In [34]:
df_14 = df_12.query("Industry == 'Hospitals' & Insurance_fulltime_proportion == 'Smaller'")
df_14 = df_14.loc[:, ['Size', 'Organization_Category', 'Insurance_offer_fulltime_employees','percent_female']]
df_14 = df_14.sort_values(by = 'percent_female', ascending= False)
df_14

Size Organization_Category Insurance_offer_fulltime_employees  \
1672  Medium            Non_profit                               Full   
1553   Large            Non_profit                            Partial   
1620   Large            Non_profit                            Partial   
1835   Small            Non_profit                               Full   
1740  Medium            Non_profit                            Partial   
1655  Medium        profit_private                               Full   
1589   Large            Non_profit                            Partial   
1659  Medium            Non_profit                               Full   
1692  Medium                   NaN                            Partial   
1705  Medium            Non_profit                            Partial   
1707  Medium            Non_profit                               Full   

      percent_female  
1672            89.0  
1553            80.0  
1620            80.0  
1835            75.0  
1740            65.0  
1655            50.0  
1589             NaN  
1659             NaN  
1692             NaN  
1705             NaN  
1707             NaN

In [35]:
myquery="""
SELECT Size, Organization_Category, Insurance_offer_fulltime_employees AS insurance, percent_female FROM whpps
WHERE Industry = 'Hospitals' AND Insurance_fulltime_proportion='Smaller'
ORDER BY percent_female DESC;
"""

In [36]:
whpps_cursor.execute(myquery)
whpps_df = whpps_cursor.fetchall()
pd.DataFrame(whpps_df)

0               1        2     3
0   Medium      Non_profit     Full  89.0
1    Large      Non_profit  Partial  80.0
2    Large      Non_profit  Partial  80.0
3    Small      Non_profit     Full  75.0
4   Medium      Non_profit  Partial  65.0
5   Medium  profit_private     Full  50.0
6    Large      Non_profit  Partial   NaN
7   Medium      Non_profit     Full   NaN
8   Medium            None  Partial   NaN
9   Medium      Non_profit  Partial   NaN
10  Medium      Non_profit     Full   NaN

## Problem 15
Write `pandas` code that replicates the output of the following SQL code:
```
SELECT industry, 
    AVG(percent_female) as percent_female, 
    AVG(percent_under30) as percent_under30,
    AVG(percent_over60) as percent_over60
FROM whpps
GROUP BY industry
ORDER BY percent_female DESC;
```
[2 points]

In [37]:
df_15 = df_12.loc[:,['Industry', 'percent_female', 'percent_under30', 'percent_over60']] 
df_15.groupby('Industry').mean().sort_values(by = 'percent_female', ascending= False)

percent_female  percent_under30  percent_over60
Industry                                                             
EducationalServices        80.657143        25.745665       11.349570
Hospitals                  76.427027        27.213793       16.489655
Arts_Entertainment         53.804416        38.566343       11.544872
Information_Finance        50.632184        23.821752       12.465465
PublicAdministration       39.056738        21.015625       15.015385
WholesaleTrade             32.657258        29.108696       12.584034
Agriculture                20.328605        22.257143       10.690355

In [38]:
myquery="""
SELECT Industry, 
    AVG(percent_female) as percent_female, 
    AVG(percent_under30) as percent_under30,
    AVG(percent_over60) as percent_over60
FROM whpps
GROUP BY Industry
ORDER BY percent_female DESC;
"""

In [39]:
whpps_cursor.execute(myquery)
whpps_df = whpps_cursor.fetchall()
pd.DataFrame(whpps_df)

0          1          2          3
0   EducationalServices  80.657143  25.745665  11.349570
1             Hospitals  76.427027  27.213793  16.489655
2    Arts_Entertainment  53.804416  38.566343  11.544872
3   Information_Finance  50.632184  23.821752  12.465465
4  PublicAdministration  39.056738  21.015625  15.015385
5        WholesaleTrade  32.657258  29.108696  12.584034
6           Agriculture  20.328605  22.257143  10.690355

## Problem 16
Write `pandas` code that replicates the output of the following SQL code:
```
SELECT gender_balance, premiums, COUNT(*)
FROM whpps
GROUP BY gender_balance, premiums
HAVING gender_balance is NOT NULL and premiums is NOT NULL;
```
[2 points]

In [40]:
df_16 = df_12.loc[:,['gender_balance', 'Insurance_offer_fulltime_employees']]
df_16.dropna(subset=['gender_balance','Insurance_offer_fulltime_employees'], inplace=True)
df_16.groupby(['gender_balance', 'Insurance_offer_fulltime_employees'])['gender_balance'].count()

gender_balance  Insurance_offer_fulltime_employees
Mostly men      Full                                  301
                None                                   91
                Partial                               332
Balanced        Full                                  226
                None                                   77
                Partial                               271
Mostly women    Full                                  267
                None                                  107
                Partial                               333
Name: gender_balance, dtype: int64

In [41]:
myquery="""
SELECT gender_balance, Insurance_offer_fulltime_employees, COUNT(*)
FROM whpps
GROUP BY gender_balance, Insurance_offer_fulltime_employees
HAVING gender_balance is NOT NULL and Insurance_offer_fulltime_employees is NOT NULL;
"""

In [42]:
whpps_cursor.execute(myquery)
whpps_df = whpps_cursor.fetchall()
pd.DataFrame(whpps_df)

0        1    2
0      Balanced     Full  226
1      Balanced     None   77
2      Balanced  Partial  271
3    Mostly men     Full  301
4    Mostly men     None   91
5    Mostly men  Partial  332
6  Mostly women     Full  267
7  Mostly women     None  107
8  Mostly women  Partial  333